### ISSUE TITLE
No compilation error when `strictNullChecks` is enabled and null check is performed on a value of  `Type | undefined`  type

### ISSUE THREAD

**Issue by pawelpabich** [Microsoft/TypeScript#21259]
**Suggestion**

We are in the process of preparing our code base to be compatible with `strictNullChecks` and noticed that certain errors in our code are not reported by TypeScript. 

**TypeScript Version:**  2.6.2

**Code**

```ts
interface Test { 
    text?: boolean; //translates to boolean | undefined
}

let variable: Test = {
    text: true
}

variable.text = undefined;

if (variable.text === null) {
    console.log("This should report an error. Shouldn't it?")
}

if (variable.text === "test") {
    console.log("This should report an error. Shouldn't it?")
}

variable.text = null; // compilation error - line 19
variable.text = "test"; // compilation error - line 20
```

```
npm run tsc -- --strictNullChecks
```

**Expected behavior:**

I would expect if checks to cause a compilation error but they don't.  This is an invalid scenario and I would like to get it reported.  

If the reasoning behind this behaviour is that there is some code that still needs to do runtime checks then I would like to have another option which would allow me to enable strict checks without any exceptions.  

**Actual behavior:**

if  checks don't cause a compilation error.

```
index.ts(19,1): error TS2322: Type 'null' is not assignable to type 'boolean | undefined'.
index.ts(20,1): error TS2322: Type '"test"' is not assignable to type 'boolean | undefined'.
```

**Playground Link:**

Removed as it did not report errors in a consistent way. 

**Related Issues:**

This might be related to https://github.com/Microsoft/TypeScript/issues/10477 but I'm not entirely sure.

**Comment by Shinigami92** [comment]
What gives you ´variable.text === null´?

**Comment by pawelpabich** [comment]
This scenario is handled correctly and it reports a compilation error.

**Comment by j-oliveras** [comment]
`variable.text == null` should not report an error because is equivalent to `variable.text === null || variable.text === undefined`.

**Comment by pawelpabich** [comment]
@j-oliveras good catch. I changed it to `===` but it still does not report if checks as errors.

**Comment by andy-ms** [comment]
There was a deliberate design decision to allow `x === null` and `x === undefined` even if `x`'s type does not include `null` or `undefined`. There is a tslint rule to check for those: https://palantir.github.io/tslint/rules/strict-type-predicates/

**Comment by pawelpabich** [comment]
Thanks @andy-ms. Is there a doco that lists this type of exceptions? It might be me but I find TS behaviour in this case surprising.

**Comment by andy-ms** [comment]
I don't know of any documentation for that, sorry.
